### DISCUSSION TITLE
Open sourcing a high-performance MRF Parser

### DISCUSSION THREAD

**Discussion by danielchalef** [CMSgov/price-transparency-guide#646] [General]
All - I recently wrote an MRF parser in Go with the objective of producing a high-performance and memory-efficient tool to be used at scale. The parser outputs a parquet dataset which can readily be ingested by other data infrastructure.

The project is available here: https://github.com/danielchalef/mrfparse

Feedback and contributions are very welcome.

-Daniel

# A Go parser for _Transparency in Coverage_ MRF files.
`mrfparse` is a memory and CPU efficient parser for _Transparency in Coverage_ Machine Readable Format (MRF) files. The parser is designed to be easily containerized and scaled on modern cloud container platforms (and potentially cloud function infrastructure).

`mrfparse` is fast: Parsing out pricing and providers for the CMS' _500 shoppable services_ from an 80GB Anthem _in-network-rates_ fileset in NDJSON format to parquet takes <5 minutes on a 12-core workstation with container memory limited to 6GB. Doing the same from the gzip compressed source file takes an additional 10 minutes.

Features:

- Outputs to a parquet dataset, allowing easy ingestion into data warehouses and data lakes.
- Supports reading from HTTP, and S3 / GS cloud storage, and writing to S3 / GS cloud storage buckets.
- Filter for a subset of CPT/HSPCS service codes (provided as a simple CSV file).
- Filters for only providers for whom pricing data is present in the MRF file, dropping extranous provider data.
- Supports reading Gzip compressed MRF files.
- The output schema is designed to support ingestion into graph databases.

**Comment by shaselton-usds** [comment]
@danielchalef Definitely cool. Thanks for sharing this. I'll take a peek and give any initial thoughts that might come to mind.
